.github/workflows/test.yml

```diff
@@ -3,13 +3,10 @@ on: [pull_request, workflow_dispatch]
 jobs:
     run_tests:
         runs-on: ubuntu-latest
-        strategy:
-            matrix:
-                node: [ '10', '12' ]
```

Comment:
At some point soon - not in this PR, I think - we should start testing on 14 alongside 12. Can we keep the `matrix` strategy and simply change the list to only include 12?

Reply:
No, I don't agree with that. If we decided to add 14 in a future PR, then the matrix strategy can be re-added at that time, but for now, YAGNI.

https://en.wikipedia.org/wiki/You_aren%27t_gonna_need_it

Reply:
I don't think this is a case where that applies. I am pretty sure that we will need it; when I say "soon" I mean like "as soon as the new upload flow is deployed". I only want to keep it in a separate, subsequent PR so that if it fails to build on 14 we can deal with those build failures there.

The reason I ask is to minimize changes: having a bunch of back-and-forth in the history makes it harder to review the history. Dropping the matrix strategy isn't the point of this change, and re-adding it won't be the point of the next change; the focus should be on the version(s) supported.

Reply:
In another [PR](https://github.com/PermanentOrg/devenv/pull/18#pullrequestreview-558243440) you wrote:

> For anyone else following along: the 10.x line is in maintenance LTS status, and leaves maintenance on 2021-04-30; the 12.x line is in active LTS status, and leaves maintenance on 2022-04-30. See Node.js § Releases. Next year we should see about upgrading to 14, but we haven't tested that on either the new upload service or on the frontend, so that's way out of scope for now.

I was unaware that upgrading to 14 was a priority. It's not been noted as a priority in any planning conversations over the past few weeks, and I can't find any open issues that referencing it. The notion that we might do something in the future is not a good enough reason to leave this in IMO.

> The reason I ask is to minimize changes: having a bunch of back-and-forth in the history makes it harder to review the history. Dropping the matrix strategy isn't the point of this change, and re-adding it won't be the point of the next change; the focus should be on the version(s) supported.

It very much is the point of this change. We were maintaining dual compatibility for the transition, and since we are doing a hard cutover, we don't need dual compatibility. I don't understand the point about making the history harder to review. `git blame` [supports revision hashes](https://git-scm.com/docs/git-blame#_specifying_ranges), the default is simply HEAD. You can also glean this kind of information from `git log`, e.g.

```
git log -L line_num_start,line_num_end:filename
```

Reply:
> I was unaware that upgrading to 14 was a priority. It's not been noted as a priority in any planning conversations over the past few weeks, and I can't find any open issues that referencing it. The notion that we might do something in the future is not a good enough reason to leave this in IMO.

Keeping up to date on our dependencies is an ongoing task. This particular dependency upgrade has a deadline: 2022-04-30. That deadline is a little ways out, but barring some existential crisis, it's not something we might do in the future, it is something we will do.

I am not saying that it's a priority, that we should do it immediately because it is important to do so. I am optimistic that this will be a low-effort upgrade, and setting up testing on the new version would either confirm or deny that theory. If it does turn out to be easy, then sure let's upgrade sooner than later, while we have recently done similar work; and if turns out not to be easy, then that's good to know so that we can plan out the work for actually addressing any potential build failures.

I'm pretty much always in favor of deleting code, but in this specific case, it's code that I already have a plan for reusing.

> It very much is the point of this change. We were maintaining dual compatibility for the transition, and since we are doing a hard cutover, we don't need dual compatibility. I don't understand the point about making the history harder to review.

Which of these is an easier diff to understand?

```diff
         runs-on: ubuntu-latest
         strategy:
             matrix:
-                node: [ '10', '12' ]
+                node: [ '12' ]
         steps:
             - uses: actions/checkout@v2
             - uses: actions/setup-node@v1
```

or

```diff
 jobs:
     run_tests:
         runs-on: ubuntu-latest
-        strategy:
-            matrix:
-                node: [ '10', '12' ]
         steps:
             - uses: actions/checkout@v2
-            - uses: actions/setup-node@v1
+            - uses: actions/setup-node@v2
               with:
-                node-version: ${{ matrix.node }}
+                node-version: "12"
             - run: npm install --production=false
             - run: npm test
```

I'm not saying that it's an unbearable hardship, but it is additional noise that a human looking at this is going to have to think through, and we can avoid that. I would prefer that we do.

Reply:
There is some good information and scoping for the node 14 upgrade in this thread. I've created an issue and linked it.

Your presumption is that we will need to maintain compatibility between node 12 and 14 when we perform the upgrade, and I don't know if we should, will or can. Given that uncertainty, I don't think it's correct to leave crufty code in the CI. However, I don't have more time to dedicate to this change. Please take over this PR, make the changes that you prefer, and get a review from a member of the Permanent team.